commands/build.go

```diff
@@ -707,3 +989,228 @@ func isExperimental() bool {
 	}
 	return false
 }
+
+func rootDataDir() (string, error) {
+	if xdh := os.Getenv("XDG_DATA_HOME"); xdh != "" {
+		return filepath.Join(xdh, "buildx"), nil
+	}
+	home := os.Getenv("HOME")
+	if home == "" {
+		return "", fmt.Errorf("environment variable HOME is not set")
+	}
+	return filepath.Join(home, ".local/share/buildx"), nil
+}
+
+type BuildxInfo struct {
+	Name    string      `json:"name"`
+	Pid     int         `json:"pid"`
+	Started time.Time   `json:"started"`
+	Control ControlInfo `json:"control"`
+}
+
+func writeBuildxInfo(dir string, info BuildxInfo) (string, error) {
+	dt, err := json.MarshalIndent(info, "", "  ")
+	if err != nil {
+		return "", err
+	}
+	f := filepath.Join(dir, "info")
+	if err := os.WriteFile(f, dt, 0600); err != nil {
+		return "", err
+	}
+	return f, nil
+}
+
+func readBuildxInfo(dir string) (*BuildxInfo, error) {
+	infoB, err := os.ReadFile(filepath.Join(dir, "info"))
+	if err != nil {
+		return nil, err
+	}
+	info := BuildxInfo{}
+	if err := json.Unmarshal(infoB, &info); err != nil {
+		return nil, err
+	}
+	return &info, nil
+}
+
+func newBuildxClientAndCheck(addr string, checkNum int, duration time.Duration) (*builder.Client, error) {
+	c, err := builder.NewClient(addr)
+	if err != nil {
+		return nil, err
+	}
+	var lastErr error
+	for i := 0; i < checkNum; i++ {
+		_, err := c.List(context.TODO())
+		if err == nil {
+			lastErr = nil
+			break
+		}
+		err = fmt.Errorf("failed to access server (tried %d times): %w", i, err)
+		logrus.Debugf("connection failure: %v", err)
+		lastErr = err
+		time.Sleep(duration)
+	}
+	if lastErr != nil {
+		return nil, lastErr
+	}
+	p, v, r, err := c.Version(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+	logrus.Debugf("connected to server (\"%v %v %v\")", p, v, r)
+	if !(p == version.Package && v == version.Version && r == version.Revision) {
+		logrus.Warnf("version mismatch (server: \"%v %v %v\", client: \"%v %v %v\"); please kill and restart buildx server",
+			p, v, r, version.Package, version.Version, version.Revision)
+	}
+	return c, nil
+}
+
+type buildxController struct {
+	*builder.Client
+	serverRoot string
+}
+
+func (c *buildxController) Kill(ctx context.Context) error {
+	info, err := readBuildxInfo(c.serverRoot)
+	if err != nil {
+		return err
+	}
+	if info.Pid <= 0 {
+		return fmt.Errorf("no PID is recorded for buildx server")
+	}
+	p, err := os.FindProcess(int(info.Pid))
+	if err != nil {
+		return err
+	}
+	if err := p.Signal(syscall.SIGINT); err != nil {
+		return err
+	}
+	// TODO: send SIGKILL if process doesn't finish.
+	return nil
+}
+
+func builderOptions2buildOption(options *builder.BuildOptions) buildOptions {
```

Comment:
Is this required? Could we remove the `builderOptions` struct entirely, and just use the public `BuildOptions` instead? For doing types like `dockeropts.MemBytes` we could do something similar to [go-digest.Digest](https://github.com/moby/buildkit/blob/master/api/services/control/control.proto#L111) in buildkit :thinking:

Reply:
Fixed to use the public `BuildOptions`.